modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/SleepTimer.kt

```diff
@@ -31,171 +30,191 @@ class SleepTimer @Inject constructor(
-        getAlarmManager().cancel(getSleepIntent())
-        cancelSleepTime()
+        updateSleepTimerStatus(sleepTimeRunning = false, sleepAfterChapters = 0, sleepAfterEpisodes = 0)
         cancelAutomaticSleepAfterTimeRestart()
         cancelAutomaticSleepOnEpisodeEndRestart()
         cancelAutomaticSleepOnChapterEndRestart()
     }
 
-    val isSleepAfterTimerRunning: Boolean
-        get() = System.currentTimeMillis() < (sleepTimeMs ?: -1)
-
-    fun timeLeftInSecs(): Int? {
-        val sleepTimeMs = sleepTimeMs ?: return null
-
-        val timeLeft = sleepTimeMs - System.currentTimeMillis()
-        if (timeLeft < 0) {
-            LogBuffer.i(TAG, "Cancelled because time is up")
-            cancelSleepTime()
-            return null
-        }
-        return (timeLeft / DateUtils.SECOND_IN_MILLIS).toInt()
+    private fun setEndOfEpisodeUuid(uuid: String) {
+        LogBuffer.i(TAG, "Episode $uuid was marked as end of episode")
+        sleepTimerHistory = sleepTimerHistory.copy(
+            lastEpisodeUuidAutomaticEnded = uuid,
+            lastTimeSleepTimeHasFinished = System.currentTimeMillis().milliseconds,
+        )
+        cancelAutomaticSleepAfterTimeRestart()
+        cancelAutomaticSleepOnChapterEndRestart()
     }
 
-    private fun getSleepIntent(): PendingIntent {
-        val intent = Intent(context, SleepTimerReceiver::class.java)
-        return PendingIntent.getBroadcast(context, 234324243, intent, PendingIntent.FLAG_IMMUTABLE)
+    private fun setEndOfChapter() {
+        LogBuffer.i(TAG, "End of chapter was reached")
+        val time = System.currentTimeMillis().milliseconds
+        sleepTimerHistory = sleepTimerHistory.copy(
+            lastSleepAfterEndOfChapterTime = time,
+            lastTimeSleepTimeHasFinished = time,
+        )
+        cancelAutomaticSleepAfterTimeRestart()
+        cancelAutomaticSleepOnEpisodeEndRestart()
     }
 
-    private fun getAlarmManager(): AlarmManager {
-        return context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
-    }
+    private fun shouldRestartSleepAfterTime(diffTime: Duration, isSleepTimerRunning: Boolean) = diffTime < MIN_TIME_TO_RESTART_SLEEP_TIMER_IN_MINUTES && sleepTimerHistory.lastSleepAfterTime != null && !isSleepTimerRunning
 
-    private fun cancelSleepTime() {
-        sleepTimeMs = null
+    private fun shouldRestartSleepEndOfEpisode(
+        diffTime: Duration,
+        currentEpisodeUuid: String,
+        isSleepEndOfEpisodeRunning: Boolean,
+    ) = diffTime < MIN_TIME_TO_RESTART_SLEEP_TIMER_IN_MINUTES && !sleepTimerHistory.lastEpisodeUuidAutomaticEnded.isNullOrEmpty() && currentEpisodeUuid != sleepTimerHistory.lastEpisodeUuidAutomaticEnded && !isSleepEndOfEpisodeRunning
+
+    private fun shouldRestartSleepEndOfChapter(diffTime: Duration, isSleepEndOfChapterRunning: Boolean) = diffTime < MIN_TIME_TO_RESTART_SLEEP_TIMER_IN_MINUTES && !isSleepEndOfChapterRunning && sleepTimerHistory.lastSleepAfterEndOfChapterTime != null
+
```

Comment:
I did not change the logic of these functions. I just moved them to the bottom of the class since they are both private